### DISCUSSION TITLE
Changing Style in Custom Control not working

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno#17628] [Q&A]
Create a simple custom Control with a Dependency Property of type `bool` and bind this property to Controls `TextBlock` which shows "on" when IsOn is `true` and "off" when IsOn is `false`:

```cs
public sealed class MyControl : Control
{
    public MyControl()
    {
        DefaultStyleKey = typeof(MyControl);
        Style = new MyStyle();
    }

    #region IsOn
    public bool IsOn
    {
        get => (bool)GetValue(IsOnProperty);
        set => SetValue(IsOnProperty, value);
    }

    public static readonly DependencyProperty IsOnProperty =
        DependencyProperty.Register(
            nameof(IsOn),
            typeof(bool),
            typeof(MyControl),
            new PropertyMetadata(false));
    #endregion
}

internal class MyStyle : Style<MyControl>
{
    public MyStyle() => Setters(s => s
        .Template(ctrl =>
            new Frame()
            .Background(Colors.LightCoral)
            .Content(new TextBlock()
                    .Text(x => x.TemplateBind(() => ctrl.IsOn).Convert<bool>(isOn => isOn ? "on" : "off"))
                    .Foreground(Colors.White)
                    .Margin(10)
                )
        ));
}
```

Use control in MainFrame:

```cs
public sealed partial class MainPage : Page
{
    public MainPage() => this
        .DataContext(new BindableMainPageModel(), (page, vm) => page
            .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
            .Content(new StackPanel()
            .VerticalAlignment(VerticalAlignment.Center)
            .HorizontalAlignment(HorizontalAlignment.Center)
            .Children(
                new TextBlock()
                .Text("Hello Uno Platform!"),
                new ToggleSwitch()
                .IsOn(x => x.Binding(() => vm.IsOn).TwoWay()),
                new MyControl()
                .Width(200)
                .Height(40)
                //.IsOn(() => vm.IsOn)                                          // works fine
                .Style(x => x.Binding(() => vm.IsOn).Convert(isOn =>            // does not work, control does not show up at all
                    isOn
                    ? new Style<MyControl>().Setters(s => s.IsOn(true))
                    : new Style<MyControl>().Setters(s => s.IsOn(false))
                    ))
            )));
}

internal partial record MainPageModel
{
    public IState<bool> IsOn => State<bool>.Value(this, () => false);
}
```

**Scenario 1**
change the value of control's `IsOn` property _directly_:

```cs
    new MyControl()
    .IsOn(() => vm.IsOn)                                          // works fine

```
Run the app on `net8.0-desktop;`: it works fine

![image](https://github.com/user-attachments/assets/a92e7a68-b560-40a9-aef2-8b33c72e6645) ![image](https://github.com/user-attachments/assets/69abda1f-9699-42b6-a6dc-6a20c1f35cb2)

**Scenario 2**
change the value of control's `IsOn` property _indirectly_ using `Style`:

```cs
    new MyControl()
    .Style(x => x.Binding(() => vm.IsOn).Convert(isOn =>            // does not work, control does not show up at all
        isOn
        ? new Style<MyControl>().Setters(s => s.IsOn(true))
        : new Style<MyControl>().Setters(s => s.IsOn(false))
        ))

```
Run the app on `net8.0-desktop;`:  it does not work, the control does not show up at all.

![image](https://github.com/user-attachments/assets/b1f6e78e-54c7-400e-bc2c-d860c7a00cba)

**Scenario 3**
Run the app as in **Scenario 1** but this time on `net8.0-windows10.0.19041;`:
 -  the control does not show up at all
 - during execution I see _Exception thrown: 'System.InvalidOperationException' in WinRT.Runtime.dll_ that is somehow handled so the app does not crash

![image](https://github.com/user-attachments/assets/f54925c9-ec3a-4e50-ac70-74b3ed0f6d14)

**Scenario 4**
Run the app as in **Scenario 2** but this time on `net8.0-windows10.0.19041;`:
 -  the control does not show up at all
 - during execution I see _Exception thrown: 'System.InvalidOperationException' in WinRT.Runtime.dll_ that is somehow handled so the app does not crash

![image](https://github.com/user-attachments/assets/f54925c9-ec3a-4e50-ac70-74b3ed0f6d14)

Note: I am using the newest dev uno branch ` "Uno.Sdk": "5.4.0-dev.145"`

repro: [UnoDynamicStyleUpdateApp.zip](https://github.com/user-attachments/files/16332651/UnoDynamicStyleUpdateApp.zip)

**Comment by MartinZikmund** [comment]
@morning4coffe-dev is doing initial investigation of this, we will then assign as needed

**Comment by kucint** [comment]
@MartinZikmund 
FYI: I forgot to remove following line from `Directory.Build.props` in my demo project. Sorry.
```
  <PropertyGroup>
    <UnoExtensionsVersion>4.1.24</UnoExtensionsVersion>
    <UnoToolkitVersion>6.0.24</UnoToolkitVersion>
    <UnoThemesVersion>5.0.13</UnoThemesVersion>
    <UnoCSharpMarkupVersion>5.2.14</UnoCSharpMarkupVersion>
  </PropertyGroup> 
```
After removing these lines, the app behaves in the same way as before, so please investigate the topic. Thanks!

**Comment by morning4coffe-dev** [comment]
Thanks for providing the repro file. After testing, it seems the issue might be that you're applying a blank Style<MyControl> to the control instead of using MyStyle. When I update the code to use MyStyle as follows:
```
.Style(x => x.Binding(() => vm.IsOn).Convert(isOn =>
isOn
? new MyStyle().Setters(s => s.IsOn(true))
: new MyStyle().Setters(s => s.IsOn(false))
))
```
I think I am seeing the desired behavior with the styles: 
![image](https://github.com/user-attachments/assets/31fb9316-d85e-42f9-932f-2cc9429e16f1)

Please let me know if this helps with your case.

**Reply by kucint** [reply]
Oh, I see!
what you are telling me, is that setting style is not set on the top of already applied style (what I was wrongly thinking) but it replaces it!
I get it now.
@morning4coffe-dev , tell me, how should I do it if I dont known which style is currently applied?
And I want to change the value of one property?

would this do the job?
```cs
.Style(x => x.Binding(() => vm.IsOn).Convert(isOn =>
isOn
? Style().Setters(s => s.IsOn(true))
: Style().Setters(s => s.IsOn(false))
))
```
or there is better way?

**Reply by MartinZikmund** [reply]
I think it is worth looping in @dansiegel here as well. I think the ideal solution would be to use `VisualStates` - I think that was also shown in https://github.com/unoplatform/uno.extensions/discussions/2319

**Reply by kucint** [reply]
right, this could be a solution in some scenarios.

**Comment by kucint** [comment]
Hi @morning4coffe-dev 
I use now `BasedOn`, but it does not work...
```cs
  .Style(x => x.Binding(() => vm.IsOn).Convert(isOn =>
      isOn
      ? new Style<MyControl>().BasedOn(Style).Setters(s => s.IsOn(true))
      : new Style<MyControl>().BasedOn(Style).Setters(s => s.IsOn(false))
      ))

```

**Reply by francoistanguay** [reply]
I would recommend for more "complex" scenarios like this that we may jump on a call and better understand what you're truly trying to accomplish.

I see that you're defining a Style (MyStyle) for MyControl that TemplateBinds the Text of TextBlock to the IsOn of MyControl and that should be valid and working, either by explicitly setting this style on the instance of MyControl or making the style implicit (used by default).

Everything else after that that you're trying to do, I dont understand and from my point of view might not be needed. So I would recommend to discuss on next sync call as it might just be some misunderstanding.

cc @MartinZikmund

**Reply by kucint** [reply]
right, let's talk about it on the next sync.
